tt_metal/impl/allocator/algorithms/free_list.cpp

```diff
@@ -46,54 +45,48 @@ std::vector<std::pair<uint64_t, uint64_t>> FreeList::available_addresses(uint64_
     return addresses;
 }
 
-FreeList::Block *FreeList::search_best(uint64_t size_bytes, bool bottom_up) {
-    FreeList::Block *best_block = nullptr;
-    FreeList::Block *curr_block = bottom_up ? this->free_block_head_ : this->free_block_tail_;
+std::shared_ptr<FreeList::Block> FreeList::search_best(uint64_t size_bytes, bool bottom_up) {
+    std::shared_ptr<FreeList::Block> best_block;
+    std::shared_ptr<FreeList::Block> curr_block = bottom_up ? this->free_block_head_ : this->free_block_tail_;
     while (curr_block != nullptr) {
         if (curr_block->size == size_bytes) {
             best_block = curr_block;
-            break;
-        } else if (curr_block->size >= size_bytes) {
-            if (best_block == nullptr or curr_block->size < best_block->size) {
-                best_block = curr_block;
-            }
+            return std::move(best_block);
```

Comment:
this now only finds free space that is exactly equal to the size to allocate. previously the "best" was either the exact size or the next largest free space. I don't think that is necessarily incorrect